edx/edx-enterprise
# [ENT-341] Change `enterprise-catalogs` & `programs` endpoints to meet requirements

**Description:** This PR modifies the initial work done for ENT-341 in the following ways:
1. Added the EnterpriseCustomerCatalog.title field.
2. Replaced the EnterpriseCustomerCatalog.query field with the EnterpriseCustomerCatalog.content_filter field. The content_filter field is a JSONField which stores query parameters that will be used to filter the search/all request made to the discovery service.
3. Removed the /enterprise/api/v1/programs/{uuid}/ endpoint.
4. Added the following endpoints:
/enterprise/api/v1/enterprise-catalogs/{uuid}/course-runs/{uuid}/
/enterprise/api/v1/enterprise-catalogs/{uuid}/programs/{uuid}/
These endpoints return the full metadata associated with the given resources instead of the condensed version returned from the search/all endpoint. Also, these endpoints will make sure the resource being requested is actually a member of the EnterpriseCustomerCatalog included in the request.
5. Modified the /enterprise/api/v1/enterprise-catalogs/{uuid}/ endpoint to return the EntepriseCustomerCatalog fields along with the paginated results of the request made to the search/all endpoint on the discovery service.
6. All requests made to any of the enterprise-catalogs endpoints will ensure that the requesting user is linked to the EnterpriseCustomerCatalog via the EnterpriseCustomerCatalog -> EnterpriseCustomer -> EnterpriseCustomerUser data model relationship.

**JIRA:** [ENT-341](https://openedx.atlassian.net/browse/ENT-341)

**Dependencies:** None.

**Merge deadline:** Asap.

**Testing instructions:**

1. Identify/create both a staff and non-staff user which you will use for testing the new endpoints.
2. [Link](https://business.sandbox.edx.org/admin/enterprise/enterprisecustomeruser/add/) both your staff and non-staff users to the PiedPiper EnterpriseCustomer.
3. [Identify/create](https://business.sandbox.edx.org/admin/enterprise/enterprisecustomercatalog/) EnterpriseCustomerCatalog models linked and unlinked with the PiedPiper EnterpriseCustomer.
4. Login as staff and access the https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs/ endpoint. Verify that all EnterpriseCustomerCatalogs are returned in the result.
5. Access the https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs/{uuid} endpoint replacing the {uuid} with one of the ECC UUIDs returned from the enterprise-catalogs endpoint. Verify that the appropriate results are returned based on the content_filter configured for the EnterpriseCustomerCatalog used.
6. Access the https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs/{ecc_uuid}/programs/{program_uuid} endpoint replacing UUIDs with appropriate UUIDs returned from the previous API calls. Verify that program detail is returned.
7. Access the https://business.sandbox.edx.org/enterprise/api/v1/enterprise-catalogs/{ecc_uuid}/course-runs/{course_run_id} endpoint replacing IDs with appropriate IDs returned from the previous API calls. Verify that course run detail is returned.
8. Repeat steps 4-7 with a non-staff user that is linked to the PiedPiper EnterpriseCustomer.
9. Repeat steps 4-7 with a non-staff user that is not-linked to the PiedPiper EnterpriseCustomer. Verify that results are NOT returned.

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.

**Post merge:**
